lib/ecto/migrator.ex

```diff
@@ -153,41 +144,65 @@ defmodule Ecto.Migrator do
   end
 
   defp do_down(repo, version, module, opts) do
-    run_maybe_in_transaction(repo, module, fn ->
+    async_migrate_maybe_in_transaction(repo, version, module, :down, opts, fn ->
       attempt(repo, version, module, :forward, :down, :down, opts)
         || attempt(repo, version, module, :backward, :change, :down, opts)
         || {:error, Ecto.MigrationError.exception(
             "#{inspect module} does not implement a `down/0` or `change/0` function")}
     end)
-    |> case do
-      :ok ->
-        verbose_schema_migration repo, "update schema migrations", fn ->
-          SchemaMigration.down(repo, version, opts[:prefix])
-        end
-        :ok
-      error ->
-        error
-    end
   end
 
-  defp run_maybe_in_transaction(repo, module, fun) do
-    fn -> do_run_maybe_in_transaction(repo, module, fun) end
-    |> Task.async()
-    |> Task.await(:infinity)
+  defp async_migrate_maybe_in_transaction(repo, version, module, direction, opts, fun) do
+    parent = self()
+    ref = make_ref()
+
+    %{pid: pid} = task =
+      Task.async(fn -> run_maybe_in_transaction(parent, ref, repo, module, fun) end)
+
+    try do
+      receive do
+        {^ref, :ok} ->
+          verbose_schema_migration repo, "update schema migrations", fn ->
+            apply(SchemaMigration, direction, [repo, version, opts[:prefix]])
+          end
+
+        {^ref, _} ->
+          :ok
+
+        {:EXIT, ^pid, _} ->
+          :ok
+      end
+    catch
+      kind, error ->
+        Task.shutdown(task, :brutal_kill)
+        :erlang.raise(kind, error, System.stacktrace())
+    else
+      _ ->
+        send(task.pid, ref)
+        Task.await(task)
```

Comment:
In the previous code the timeout was set to `:infinity`. My first guess is to support migrations that last longer than 5 seconds, should it be set here too?

Reply:
Yes, excellent catch.